MaxText/layers/moe.py

```diff
@@ -357,29 +360,152 @@ def unpermute(self, intermediate, sorted_selected_experts, weights, batch_size,
       )
     return output.reshape(batch_size, sequence_length, -1).astype(self.dtype)
 
-  def sparse_matmul(self, inputs, gate_logits, pre_bias_logits, w0_kernel, w1_kernel, wo_kernel):
-    """sparse matrix multiplication"""
-    tile_size = (512, 1024, 1024)  # (m, k, n)
-
-    def local_permute(inputs, global_group_sizes, local_expert_size):
-      """Sort inputs by expert within each shard."""
-      local_id = jax.lax.axis_index("expert")
-      local_sizes = jax.lax.dynamic_slice_in_dim(
-          global_group_sizes, local_id * local_expert_size, local_expert_size, axis=1
-      ).reshape(-1)
+  @staticmethod
+  def local_permute(inputs, global_group_sizes, local_expert_size, shard_index, is_offset=False, global_sorted_experts=None):
+    """Permutes tokens locally within an expert shard.
+
+    This function prepares the input tokens for processing by the experts located
+    on the current shard. It groups the tokens by their assigned local expert
+    index (0 to local_expert_size - 1).
+
+    Args:
+      inputs: The input data (tokens) assigned to the experts on this shard.
+        Shape `[tokens, emb_dim]`.
+      global_group_sizes: The count of tokens assignments for each global expert across all the batch shards.
+        Shape `[num_batch_shards, num_experts].
+      local_expert_size: The number of experts handled by the current shard.
+      shard_index: The index of the current expert shard (0 to num_expert_parallelism - 1).
+      is_offset: If True, assumes `inputs` are pre-sorted by global expert ID
+        and selects the slice relevant to this shard's assigned experts. If False, assumes
+        that `inputs` corresponding to the shard's experts start from the beginning of the tensor
+        but need to be permuted by expert ID.
+      global_sorted_experts: Global expert IDs for the `inputs` used when `is_offset`
+        is True. Shape `[total_tokens_for_this_shard]`.
+
+    Returns:
+      A tuple containing:
+        sorted_inputs: Input data permuted local expert ID.
+        sorted_indices: Indices used to permute the inputs.
+        local_group_size: Number of tokens assigned to each local expert on this
+          shard.
+        sorted_experts_ids: expert ID corrsponding to each token of the permuted inputs.
+    """
+
+    # Slice the count of local expert IDs in each batch shard.
+    all_shard_local_sizes = jax.lax.dynamic_slice_in_dim(
```

Comment:
When rebase from head, could you help keep those shape info for some variables? I think I had some updates there. Thank you!